### PR TITLE
[SPARK-16904] [SQL] Removal of Hive Built-in Hash Functions and TestHiveFunctionRegistry

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?

Currently, the Hive built-in `hash` function is not being used in Spark since Spark 2.0. The public interface does not allow users to unregister the Spark built-in functions. Thus, users will never use Hive's built-in `hash` function. 

The only exception here is `TestHiveFunctionRegistry`, which allows users to unregister the built-in functions. Thus, we can load Hive's hash function in the test cases. If we disable it, 10+ test cases will fail because the results are different from the Hive golden answer files.

This PR is to remove `hash` from the list of `hiveFunctions` in `HiveSessionCatalog`. It will also remove `TestHiveFunctionRegistry`. This removal makes us easier to remove `TestHiveSessionState` in the future.
### How was this patch tested?
N/A

### CODE REVIEW

**Comment by gatorsmile** [comment]
retest this please

**Comment by hvanhovell** [comment]
This seems fair. Shouldn't we do something about all the disabled test cases though?

**Comment by gatorsmile** [comment]
Yeah, we should check all the disabled test cases and see if we can move it. I am waiting for https://github.com/apache/spark/pull/14472. : )

**Comment by gatorsmile** [comment]
#14472 has been merged. Will work on it soon.

**Comment by tejasapatil** [comment]
Is Spark's hashing function semantically equivalent to Hive's ? AFAIK, its not. I think it would be better to have a mode to be able to use Hive's hash method. eg. case when this would be needed: Users running a query in Hive want to switch to Spark. As this happens, you want to verify if the data produced is same or not. Also, for a brief time the pipeline would run in both the engines. Upstream consumers of the data generated should not see differences due to running in the different engines

**Comment by gatorsmile** [comment]
@tejasapatil Thanks for the inputs! I can understand your usage cases. It looks very reasonable. 

Actually, in Spark 2.0, we already removed it. You are unable to load/invoke the Hive's hash function. This PR is just to clean the usage of `hash` function in our testing package.

**Comment by rxin** [comment]
@gatorsmile can we take this to the finish line? This actually does not impact hash at all since it's only here for testing.

**Comment by gatorsmile** [comment]
@rxin Sure, will do it this weekend. Thanks!

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala [hunk 1]
```diff
@@ -487,24 +487,6 @@ private[hive] class TestHiveQueryExecution(
   }
 }
 
-
-private[hive] class TestHiveFunctionRegistry extends SimpleFunctionRegistry {
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala]: We can still remove this class if we add back the removed spark builtin hash function manually right?

Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala]: Yeah, I think so

**Comment by rxin** [comment]
If this is no longer WIP, please update the title. Thanks.

**Comment by gatorsmile** [comment]
Based on the previous discussions in the other PRs, it sounds like these Hive-specific test cases are not very useful. Do we still need to port them back? 
```
    "auto_join19",
    "auto_join22",
    "auto_join25",
    "auto_join26",
    "auto_join27",
    "auto_join28",
    "auto_join30",
    "auto_join31",
    "auto_join_nulls",
    "auto_join_reordering_values",
    "correlationoptimizer1",
    "correlationoptimizer2",
    "correlationoptimizer3",
    "correlationoptimizer4",
    "multiMapJoin1",
    "orc_dictionary_threshold",
    "udf_hash"
```

**Comment by rxin** [comment]
I'm ok with not having them.

**Comment by rxin** [comment]
Merging in master/branch-2.1.

**Pull request closed without merging**